src/preprocess/annotator_base.py

```diff
@@ -18,15 +18,23 @@ def __init__(self, preprocessor: Preprocessor, use_cache=True):
 
     @staticmethod
     def _par_sample_train_data(marked_doc):
+        print("Processing sents with POS and NEG samples - base")
         sents = marked_doc['sents']
+        # Here, we should insert the negative keyphrases such as `product solving`
         for sent in sents:
             phrases = sent['phrases']
             assert phrases
             positive_spans = [tuple(phrase[0]) for phrase in phrases]
             num_positive = len(positive_spans)
             # sample negatives
             word_idxs = sent['widxs']
+            # Iterave over all ngrams of length MINGRAMS to MAXGRAMS
+            # (0, 4), (1, 5)...
+            # (0, 3), (1, 4), ...
+            # (0, 2), (1, 3), ...
+            # (0, 1), (1, 2), ...
             all_spans = utils.get_possible_spans(word_idxs, len(sent['ids']), consts.MAX_WORD_GRAM, consts.MAX_SUBWORD_GRAM)
+            # Discard positive spans in negative_spans
```

Comment:
does this mean, if there is a positive sample in a negative sample, we discard the positive sample?

Reply:
No, this part simply generates the negative samples. So the steps are:
- Create all possible spans in a variable called `all_spans`
- Remove the **positive spans** from the `all_spans` to not consider positive spans as negative.
- Select randomly negative spans from `all_spans`